packages/data/src/index.js

```diff
@@ -280,44 +279,46 @@ export const withSelect = ( mapStateToProps ) => createHigherOrderComponent( ( W
 			 * @type {boolean}
 			 */
 			this.shouldComponentUpdate = false;
+			this.state = {
+				mergeProps: mapStateToProps( select, props ) || {},
```

Comment:
That said, it could be a micro-optimization since strict equality is a shortcutting case in `@wordpress/is-shallow-equal`:

https://github.com/WordPress/packages/blob/c24da3afe3fa4b1ccd234f8c2ec766d43201c1ce/packages/is-shallow-equal/objects.js#L16-L18